pkg/envoy/server/cache.go

```diff
@@ -0,0 +1,96 @@
+package ads
+
+import (
+	"context"
+	"errors"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/rs/zerolog"
+)
+
+// OnStreamOpen is called on stream open
+func (s *Server) OnStreamOpen(ctx context.Context, streamID int64, typ string) error {
+	log.Debug().Msgf("OnStreamOpen id: %d typ: %s", streamID, typ)
+	return s.callbacks.ProxyConnected(ctx, streamID)
+}
+
+// OnStreamClosed is called on stream closed
+func (s *Server) OnStreamClosed(streamID int64) {
+	log.Debug().Msgf("OnStreamClosed id: %d", streamID)
+	s.callbacks.ProxyDisconnected(streamID)
+}
+
+// OnStreamRequest is called when a request happens on an open connection
+func (s *Server) OnStreamRequest(streamID int64, req *discovery.DiscoveryRequest) error {
+	log.Debug().Msgf("OnStreamRequest node: %s, type: %s, v: %s, nonce: %s, resNames: %s", req.Node.Id, req.TypeUrl, req.VersionInfo, req.ResponseNonce, req.ResourceNames)
+	return nil
+}
+
+// OnStreamResponse is called when a response is being sent to a request
+func (s *Server) OnStreamResponse(_ context.Context, streamID int64, req *discovery.DiscoveryRequest, resp *discovery.DiscoveryResponse) {
+	log.Debug().Msgf("OnStreamDeltaResponse node: %s type: %s, v: %s, nonce: %s, NumResources: %d", req.Node.Id, resp.TypeUrl, resp.VersionInfo, resp.Nonce, len(resp.Resources))
+}
+
+// --- Fetch request types. Callback interfaces still requires these to be defined
+
+// OnFetchRequest is called when a fetch request is received
+func (s *Server) OnFetchRequest(_ context.Context, req *discovery.DiscoveryRequest) error {
+	// Unimplemented
+	return errors.New("Unsupported XDS server connection type")
+}
+
+// OnFetchResponse is called when a fetch request is being responded to
+func (s *Server) OnFetchResponse(req *discovery.DiscoveryRequest, resp *discovery.DiscoveryResponse) {
+	// Unimplemented
```

Comment:
I'd prefer an explicit error here as well for consistency's sake

Reply:
this has to fit the interface, so an error can't be returned. Note that this won't ever be called since the OnFetchRequest will preempt this